src/client/auth.js

```diff
@@ -1,60 +1,146 @@
 const UUID = require('uuid-1345')
 const yggdrasil = require('yggdrasil')
+const fs = require('fs').promises
 
 module.exports = function (client, options) {
-  const yggdrasilClient = yggdrasil({ agent: options.agent, host: options.authServer || 'https://authserver.mojang.com' })
-  const clientToken = options.clientToken || (options.session && options.session.clientToken) || UUID.v4().toString()
-  const skipValidation = false || options.skipValidation
-  options.accessToken = null
-  options.haveCredentials = options.password != null || (clientToken != null && options.session != null)
-
-  if (options.haveCredentials) {
-    // make a request to get the case-correct username before connecting.
-    const cb = function (err, session) {
-      if (err) {
-        client.emit('error', err)
-      } else {
-        client.session = session
-        client.username = session.selectedProfile.name
-        options.accessToken = session.accessToken
-        client.emit('session', session)
-        options.connect(client)
+  (async function() {
```

Comment:
Can't you just make the exported function async?

Reply:
See https://github.com/PrismarineJS/node-minecraft-protocol/pull/780#discussion_r533308810
I think we can make this async but for me we need to handle the promise rejection somewhere, so if this is made async I think we should handle the error in createClient.js

Reply:
It was like this before, I wasn't sure if that was okay

Reply:
Of course we have to handle it. 

If things haven't changed (I've been busy lately), we're using callbacks in pretty much every project so another option is to transform this async function into a callback one at some point.

The rejection can also emit an error of some sort.